### PR TITLE
Add uneven load training feature

### PR DESCRIPTION
**Pull request by XericZephyr**
I've created a uneven load training mechanism for multiple devices in a single machine. 

## Usage 

Here is my test code modified from `example/imagenet/inception.py`
```python
work_load_list = [(0, 5), (1, 1.5), (2, 5)]
gpus = [mx.gpu(i, work_load) for i, work_load in work_load_list]

model = mx.model.FeedForward(ctx=gpus, symbol=softmax, num_epoch=num_epoch,
                             learning_rate=0.05, momentum=0.9, wd=0.00001)

model.fit(X=train, eval_data=val,
          eval_metric="acc",
          batch_end_callback=mx.callback.Speedometer(batch_size),
          epoch_end_callback=mx.callback.do_checkpoint(model_prefix))
```

Here, `work_load_list` is a list of `(device_index, work_load)` tuple. Pass `gpus` to the `ctx` argument for `model.fit` API could automatically unevenly distribute samples to different devices. 

**If you don't want this feature, you're good to use your old code with this new commit.**

## Result

My test machine is a workstation equipped with a Quadro K4000 and two Tesla K40c. Training with the modified `example/imagenet/inception.py`, with ratio `10:10:3` for `K40c:K40c:K4000`, the training speedometer is shown as: 

```
INFO:root:Iter[0] Batch [100]	Speed: 92.63 samples/sec
INFO:root:Iter[0] Batch [150]	Speed: 91.22 samples/sec
INFO:root:Iter[0] Batch [200]	Speed: 92.20 samples/sec
INFO:root:Iter[0] Batch [250]	Speed: 92.53 samples/sec
INFO:root:Iter[0] Batch [300]	Speed: 92.55 samples/sec
```

With only two Tesla GPUs, we observe a speed around 

```
INFO:root:Iter[0] Batch [12700]	Speed: 80.39 samples/sec
INFO:root:Iter[0] Batch [12750]	Speed: 80.27 samples/sec
INFO:root:Iter[0] Batch [12800]	Speed: 80.31 samples/sec
```

So this uneven work load feature is able to squeeze the last GPU on the machine and slightly improve the performance.

### CODE REVIEW

**Comment by tqchen** [comment]
hmm, I think we want to debate whether it is good to put the work_load into the context class, or simply add it as a optional training parameter.

Also it might be bad to have a small batch for batchnormed network

**Comment by XericZephyr** [comment]
Yes. I've also considered this. Adding the `work_load` into Context class might have influenced the the **abstract** of Context concept, since it's no use for data placement management. It's only used for training. 

However, I think the code in usage is more elegant for users to associate the `work_load` with the specific device, which is why I add the `work_load` to Context **abstract**.

**Comment by mli** [comment]
many thanks for the pr. the results are quite nice. 

i also suggest to add another argument beside `ctx` when creating the model rather than change the context, since it makes the python inconsistent with other lang bindings. adding the work_load to a device is nature, but it cannot extend to distributed training, where we need to manually partition the data.

btw, is there a way to query the gpu computational power so we can decide it automatically?

**Comment by XericZephyr** [comment]
Received. I will remove the hack from Context and add an additional argument to solve the uneven load problem. 

For automatically query, I think it would be very complicated, since there's not a general formula to calculate the overall computational power of a GPU. Plus sometimes we also need to consider the usage of GPU memory. 

However, if you have different GPUs, I would suggest the following strategy to adjust your `work_load` assignment. 
1. `watch -n 1 nvidia-smi` to closely watch the usage of your GPUs. 
2. If you observe some GPU's `Volatile GPU-Util` periodically drop to 0% and then goes back to 99%. That means this GPU is waiting others to calculate and synchronize. So you need to add some work load to that GPU. 
3. If all GPU keep `99%` and steady, congratulations, you squeeze all the performance of your GPUs.

**Comment by XericZephyr** [comment]
I recover the `Context` abstract to its origin and add the `work_load_list` argument to both `fit` and `create` function. For now, the usage is 

```python
num_gpu = 3
gpus = [mx.gpu(i) for i in range(num_gpu)]
work_load_list = [10, 3, 10]

model = mx.model.FeedForward(ctx=gpus, symbol=softmax, num_epoch=num_epoch,
                             learning_rate=0.05, momentum=0.9, wd=0.00001)

model.fit(X=train, eval_data=val,
          eval_metric="acc",
          batch_end_callback=mx.callback.Speedometer(batch_size),
          epoch_end_callback=mx.callback.do_checkpoint(model_prefix),
          work_load_list=work_load_list)
```

**Comment by piiswrong** [comment]
I think a simple way of doing automatic query would be just run training loop for a few iterations and time it.

**Comment by XericZephyr** [comment]
@piiswrong  I think it's a reasonable choice. Can we add GPU memory limit into consideration?  Since usually we want to set `batch_size` as large as possible to fit into the GPU memory.

**Comment by XericZephyr** [comment]
For now, I think it's probably better to leave the user freedom to split the `work_load` to different devices.  I don't think the automatic query could always give an optimized performance.

For uneven load feature, I think it's ready to merge now. 

BTW, could we find a way to eliminate the `too-many-arguments` in the `fit` and `create` function in `model.py` ?

**Comment by tqchen** [comment]
Thanks for the contribution, this is now merged, have added your name to https://github.com/dmlc/mxnet/blob/master/CONTRIBUTORS.md , please feel free to update your information or contribution more to mxnet!
